extensions-core/druid-lookups/src/main/java/io/druid/server/lookup/PollingLookup.java

```diff
@@ -0,0 +1,259 @@
+      refOfCacheKeeper.set(new CacheRefKeeper(this.cacheFactory.makeOf(dataFetcher.fetchAll())));
+      if (getPollMs() > 0) {
+        pollFuture = scheduledExecutorService.scheduleAtFixedRate(
+            pollAndSwap(),
+            getPollMs(),
+            getPollMs(),
+            TimeUnit.MILLISECONDS
+        );
+      } else {
+        pollFuture = null;
+      }
+      this.isOpen.set(true);
+    }
+  }
+
+
+  public void close()
+  {
+    LOGGER.info("Closing lookup");
+    synchronized (isOpen) {
+      isOpen.getAndSet(false);
+      CacheRefKeeper cacheRefKeeper = refOfCacheKeeper.getAndSet(null);
+      if (cacheRefKeeper != null) {
+        cacheRefKeeper.doneWithIt();
+      }
+      if (pollFuture != null) {
+        pollFuture.cancel(true);
+      }
+      scheduledExecutorService.shutdown();
+    }
+  }
+
+  @Override
+  public String apply(@NotNull String key)
+  {
+    assertIsOpen();
+    CacheRefKeeper cacheRefKeeper = refOfCacheKeeper.get();
+    if (cacheRefKeeper == null) {
+      throw new ISE("Cache reference is null WTF");
+    }
+    PollingCache cache = cacheRefKeeper.get();
+    try {
+      if (cache == null) {
+        // it must've been closed while I was getting it.  Try again.
+        return this.apply(key);
+      }
+      return Strings.emptyToNull((String) cache.get(key));
+    }
+    finally {
+      if (cacheRefKeeper != null && cache != null) {
+        cacheRefKeeper.doneWithIt();
+      }
+    }
+  }
+
+  @Override
+  public List<String> unapply(final String value)
+  {
+    assertIsOpen();
+    CacheRefKeeper cacheRefKeeper = refOfCacheKeeper.get();
+    if (cacheRefKeeper == null) {
+      throw new ISE("lookup is closed");
+    }
+    PollingCache cache = cacheRefKeeper.get();
+    try {
+      if (cache == null) {
+        // it must've been closed while I was getting it.  Try again.
+        return this.unapply(value);
+      }
+      return cache.getKeys(value);
+    }
+    finally {
+      if (cacheRefKeeper != null && cache != null) {
+        cacheRefKeeper.doneWithIt();
+      }
+    }
+  }
+
+  @Override
+  public byte[] getCacheKey()
+  {
+    assertIsOpen();
+    byte[] randomKey = StringUtils.toUtf8(UUID.randomUUID().toString());
+    ByteBuffer byteArray = ByteBuffer.allocate(1 + randomKey.length);
+    return byteArray.put(CACHE_ID).put(randomKey).array();
+  }
+
+  private Runnable pollAndSwap()
+  {
+    return new Runnable()
+    {
+      @Override
+      public void run()
+      {
+        LOGGER.debug("Polling and swapping of lookup");
+        CacheRefKeeper newCacheKeeper = new CacheRefKeeper(cacheFactory.makeOf(dataFetcher.fetchAll()));
+        CacheRefKeeper oldCacheKeeper = refOfCacheKeeper.getAndSet(newCacheKeeper);
+        if (oldCacheKeeper != null) {
+          oldCacheKeeper.doneWithIt();
+        }
+      }
+    };
+  }
+
+  private long getPollMs()
+  {
+    return pollPeriod.toStandardDuration().getMillis();
+  }
+
+  private void assertIsOpen() throws ISE
```

Comment:
throws ISE is not needed, usually you would not declare RuntimeException ... if it needs to be declared then it should really be a checked exception.